{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# %% [markdown]\n",
    "# ## 1. Processamento do Arquivo GFF\n",
    "# Extrair coordenadas genômicas (start, end) para os genes filtrados"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total de genes no GFF: 3362\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>Prodigal:002006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>330.0</td>\n",
       "      <td>1175.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=JDFEPHJD_00001;inference=ab initio predicti...</td>\n",
       "      <td>JDFEPHJD_00001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>Prodigal:002006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>2326.0</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=JDFEPHJD_00002;eC_number=1.1.1.86;Name=ilvC...</td>\n",
       "      <td>JDFEPHJD_00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>Prodigal:002006</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2494.0</td>\n",
       "      <td>3375.0</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=JDFEPHJD_00003;Name=oxyR_1;gene=oxyR_1;infe...</td>\n",
       "      <td>JDFEPHJD_00003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                seqid           source type   start     end  \\\n",
       "0  NODE_1_length_1650131_cov_5.385719  Prodigal:002006  CDS   330.0  1175.0   \n",
       "1  NODE_1_length_1650131_cov_5.385719  Prodigal:002006  CDS  1310.0  2326.0   \n",
       "2  NODE_1_length_1650131_cov_5.385719  Prodigal:002006  CDS  2494.0  3375.0   \n",
       "\n",
       "  score strand phase                                         attributes  \\\n",
       "0     .      -     0  ID=JDFEPHJD_00001;inference=ab initio predicti...   \n",
       "1     .      -     0  ID=JDFEPHJD_00002;eC_number=1.1.1.86;Name=ilvC...   \n",
       "2     .      +     0  ID=JDFEPHJD_00003;Name=oxyR_1;gene=oxyR_1;infe...   \n",
       "\n",
       "          gene_id  \n",
       "0  JDFEPHJD_00001  \n",
       "1  JDFEPHJD_00002  \n",
       "2  JDFEPHJD_00003  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%\n",
    "import pandas as pd\n",
    "\n",
    "# Carregar o arquivo GFF (substitua pelo seu caminho)\n",
    "gff_path = \"/Users/rvincenzi/Documents/USP/Doutorado/Tese/genomas_geral/genomica_comparativa/eggnog/p4d.fna.gff\"\n",
    "gff_cols = [\"seqid\", \"source\", \"type\", \"start\", \"end\", \"score\", \"strand\", \"phase\", \"attributes\"]\n",
    "gff_df = pd.read_csv(gff_path, sep=\"\\t\", comment=\"#\", header=None, names=gff_cols)\n",
    "\n",
    "# Filtrar apenas genes (ajuste 'type' conforme seu GFF)\n",
    "gene_df = gff_df[gff_df[\"type\"].isin([\"gene\", \"CDS\"])].copy()\n",
    "\n",
    "# Extrair o ID do gene dos atributos (ex: \"ID=gene0001;Name=kdpA\")\n",
    "gene_df[\"gene_id\"] = gene_df[\"attributes\"].str.extract(r\"ID=([^;]+)\")\n",
    "\n",
    "# Mostrar estrutura\n",
    "print(f\"Total de genes no GFF: {len(gene_df)}\")\n",
    "gene_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# %% [markdown]\n",
    "# ## 2. Cruzamento com os Genes de Interesse\n",
    "# Vincular as coordenadas aos genes salt-in/salt-out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Genes sem coordenadas no GFF: 0\n"
     ]
    }
   ],
   "source": [
    "# %%\n",
    "# Carregar sua lista filtrada (resultado do passo anterior)\n",
    "filtered_genes = pd.read_csv(\"p4d_genes_salt_filtrados.csv\")\n",
    "\n",
    "# Juntar com o GFF usando o ID do gene\n",
    "merged_df = pd.merge(\n",
    "    filtered_genes,\n",
    "    gene_df[[\"gene_id\", \"seqid\", \"start\", \"end\", \"strand\"]],\n",
    "    left_on=\"query\",  # Coluna com IDs no seu arquivo filtrado\n",
    "    right_on=\"gene_id\",\n",
    "    how=\"left\"\n",
    ")\n",
    "\n",
    "# Verificar genes sem coordenadas\n",
    "missing_coords = merged_df[merged_df[\"start\"].isna()]\n",
    "print(f\"\\nGenes sem coordenadas no GFF: {len(missing_coords)}\")\n",
    "if len(missing_coords) > 0:\n",
    "    print(missing_coords[[\"query\", \"Preferred_name\"]].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Arquivo gerado\n",
      "Exemplo das primeiras linhas:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqid</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>color</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JDFEPHJD_00029</td>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>30539.0</td>\n",
       "      <td>31675.0</td>\n",
       "      <td>-</td>\n",
       "      <td>#e74c3c</td>\n",
       "      <td>proB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JDFEPHJD_00399</td>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>431809.0</td>\n",
       "      <td>433095.0</td>\n",
       "      <td>-</td>\n",
       "      <td>#e74c3c</td>\n",
       "      <td>proA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JDFEPHJD_00429</td>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>461140.0</td>\n",
       "      <td>461748.0</td>\n",
       "      <td>+</td>\n",
       "      <td>#e74c3c</td>\n",
       "      <td>betI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JDFEPHJD_00430</td>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>461790.0</td>\n",
       "      <td>463274.0</td>\n",
       "      <td>+</td>\n",
       "      <td>#e74c3c</td>\n",
       "      <td>betB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JDFEPHJD_00999</td>\n",
       "      <td>NODE_1_length_1650131_cov_5.385719</td>\n",
       "      <td>1134976.0</td>\n",
       "      <td>1137774.0</td>\n",
       "      <td>+</td>\n",
       "      <td>#3498db</td>\n",
       "      <td>nhAB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id                               seqid      start        end  \\\n",
       "0  JDFEPHJD_00029  NODE_1_length_1650131_cov_5.385719    30539.0    31675.0   \n",
       "1  JDFEPHJD_00399  NODE_1_length_1650131_cov_5.385719   431809.0   433095.0   \n",
       "2  JDFEPHJD_00429  NODE_1_length_1650131_cov_5.385719   461140.0   461748.0   \n",
       "3  JDFEPHJD_00430  NODE_1_length_1650131_cov_5.385719   461790.0   463274.0   \n",
       "4  JDFEPHJD_00999  NODE_1_length_1650131_cov_5.385719  1134976.0  1137774.0   \n",
       "\n",
       "  strand    color label  \n",
       "0      -  #e74c3c  proB  \n",
       "1      -  #e74c3c  proA  \n",
       "2      +  #e74c3c  betI  \n",
       "3      +  #e74c3c  betB  \n",
       "4      +  #3498db  nhAB  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# %%\n",
    "# Criar arquivo no formato esperado pelo Proksee\n",
    "proksee_df = merged_df.dropna(subset=[\"start\"]).copy()\n",
    "proksee_df[\"color\"] = proksee_df[\"strategy\"].map({\n",
    "    \"salt-in\": \"#3498db\",  # Azul\n",
    "    \"salt-out\": \"#e74c3c\"  # Vermelho\n",
    "})\n",
    "proksee_df[\"label\"] = proksee_df[\"gene_found\"]  # Ou use \"Preferred_name\"\n",
    "\n",
    "# Selecionar colunas e salvar\n",
    "proksee_export = proksee_df[[\"gene_id\", \"seqid\", \"start\", \"end\", \"strand\", \"color\", \"label\"]]\n",
    "proksee_export.to_csv(\"p4d_proksee_salt_genes.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "print(f\"✅ Arquivo gerado\")\n",
    "print(\"Exemplo das primeiras linhas:\")\n",
    "proksee_export.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
